compiler/rustc_mir_transform/src/coverage/counters.rs

```diff
@@ -181,6 +181,8 @@ impl CoverageCounters {
     pub(super) fn bcb_edge_counters(
         &self,
     ) -> impl Iterator<Item = (BasicCoverageBlock, BasicCoverageBlock, &BcbCounter)> {
+        // The contract of bcb_edge_counters() makes no guarantees about ordering.
```

Comment:
Do the use sites respect this contract ?

Reply:
The only use site seems to be

https://github.com/rust-lang/rust/blob/be6b059169aa8e7cde3a80fdadd772a4b77f55ff/compiler/rustc_mir_transform/src/coverage/mod.rs#L169-L194

and it looks to me to be about inserting counters for code coverage, which intuitively seems fine to do in any order.

Disclaimer: I'm not very familiar with these parts of the compiler, so my analysis might be flawed.

Reply:
Cc @Zalathar for your expertise. 

IIUC, the use site injects counters, so modifies MIR, which is the query's result. So we need to ensure determinism here.

Reply:
The use site is *intended* to be order-insensitive, but because the code creates a new block for each edge counter, it might accidentally be non-deterministic.

(The resulting MIR would be equivalent but not necessarily identical, because the individual blocks might be created in a different order and thus have different IDs relative to each other.)

Switching over to an ordered hashmap in `counters.rs` should make it deterministic. I’ll keep looking into this to see if I spot any other concerns.

Reply:
I filed #119401 to switch over to using `FxIndexMap` here, which should make this `#[allow(...)]` unnecessary.

Feel free to cherry-pick it into this PR if you want.

Reply:
Thanks! My preference would be to land your PR first and that I then rebase this PR.

If you prefer me to cherry-pick, I can do that too, but it might take a while until I have time to investigate the remaining open concerns.

Reply:
My PR has been merged, so when you have a chance to rebase you should be able to get rid of this exemption.